src/libraries/System.Private.CoreLib/src/System/DateTime.Windows.cs

```diff
@@ -1,30 +1,164 @@
+                // constructing a new DateTime object from these components, but this is slow.
+                // So instead we'll rely on the fact that leap seconds only ever adjust the day
+                // by +1 or -1 second, and only at the very end of the day. That is, time rolls
+                // 23:59:58 -> 00:00:00 (negative leap second) or 23:59:59 -> 23:59:60 (positive leap
+                // second). Thus we assume that each day has at least 23 hr 59 min 59 sec, or
+                // 863,990,000,000 ticks.
+                //
+                // We take advantage of this by caching what the OS believes the tick count is at
+                // the beginning of the month vs. what .NET believes the tick count is at the beginning
+                // of the month. When the OS returns a tick count to us, if it's before 23:59:59 on the
+                // last day of the month, then we know that there's no way for a leap second to have been
+                // inserted or removed, and we can short-circuit the leap second handling logic by
+                // performing a quick addition and returning immediately.
+                //
+                // This relies on:
+                // a) Leap seconds only ever taking place on the last day of the month
+                //    (historically these are Jun. & Dec., but we support any month);
+                // b) Leap seconds only ever occurring at 23:59:59;
+                // c) At most a single leap second being added or removed per month; and
+                // d) Windows never inserting a historical leap second (into the past) once the
+                //    process is up and running (future insertions are ok).
+
+                // TODO: Once https://github.com/dotnet/runtime/issues/38134 is resolved,
+                // the call below should suppress the GC transition.
+
+                ulong osTicks;
+                s_pfnGetSystemTimeAsFileTime(&osTicks);
+
+                // If the OS doesn't support leap second handling, short-circuit everything.
+
+                if (!s_systemSupportsLeapSeconds)
                 {
-                    FullSystemTime time;
-                    GetSystemTimeWithLeapSecondsHandling(&time);
-                    return CreateDateTimeFromSystemTime(in time);
+                    return new DateTime((osTicks + FileTimeOffset) | KindUtc);
                 }
 
-                return new DateTime(((ulong)(GetSystemTimeAsFileTime() + FileTimeOffset)) | KindUtc);
+                // If our cache has been populated and we're within its validity period, rely
+                // solely on the cache. We use unsigned arithmetic below so that if the system
+                // clock is set backward, the resulting value will integer underflow and cause
+                // a cache miss, forcing us down the slow path.
+
+                if (s_leapSecondCache is LeapSecondCache cache)
```

Comment:
Do we need to do this check every time too? can we just initialize instance with some value? or you have it for optimization too?

Reply:
Interesting idea. We could initialize it with a dummy value like (0, 0, 0). It would save two instructions (a comparison + conditional branch) in the x64 codegen. Benchmark doesn't really show a difference since the branch predictor is primed to assume the field is initialized.

|    Method |        Job |       Toolchain |     Mean |    Error |   StdDev | Ratio |
|---------- |----------- |---------------- |---------:|---------:|---------:|------:|
| GetUtcNow | Job-LSFSOO |           proto | 28.97 ns | 0.112 ns | 0.104 ns |  1.00 |
| GetUtcNow | Job-LKFIYO | proto_initcache | 28.85 ns | 0.186 ns | 0.174 ns |  1.00 |

Even if there's no real perf difference, might be worth doing anyway just to simplify things?